clang-tools-extra/clang-tidy/daedalean/IncludeOrderCheck.cpp

```diff
@@ -0,0 +1,183 @@
+                          const Module *Imported,
+                          SrcMgr::CharacteristicKind FileType) override;
+  void EndOfMainFile() override;
+
+private:
+  struct IncludeDirective {
+    SourceLocation Loc;
+    CharSourceRange Range;
+    std::string Filename;
+
+    GroupType groupType;
+    std::string project;
+
+    bool operator <(const IncludeDirective & other) const {
+      if (groupType != other.groupType) {
+        return static_cast<int>(groupType) < static_cast<int>(other.groupType);
+      }
+      return  Filename < other.Filename;
+    }
+  };
+
+  typedef std::vector<IncludeDirective> FileIncludes;
+  std::map<clang::FileID, FileIncludes> IncludeDirectives;
+
+  ClangTidyCheck &Check;
+  const SourceManager &SM;
+
+private:
+  void printInverse(const IncludeDirective & what, const IncludeDirective & beforeWhat);
+};
+} // namespace
+
+void IncludeOrderCheck::registerPPCallbacks(const SourceManager &SM,
+                                            Preprocessor *PP,
+                                            Preprocessor *ModuleExpanderPP) {
+  PP->addPPCallbacks(::std::make_unique<IncludeOrderPPCallbacks>(*this, SM));
+}
+
+void IncludeOrderPPCallbacks::InclusionDirective(
+    SourceLocation HashLoc, const Token &IncludeTok, StringRef FileName,
+    bool IsAngled, CharSourceRange FilenameRange, const FileEntry *File,
+    StringRef SearchPath, StringRef RelativePath, const Module *Imported,
+    SrcMgr::CharacteristicKind FileType) {
+
+  if (!File) {
+    Check.diag(FilenameRange.getBegin(), "Include <" + FileName.str() + "> not resolved");
+    return;
+  }
+
+  if (!IsAngled) {
+    Check.diag(FilenameRange.getBegin(), "Use #include<> instead of include \"\"") << FixItHint::CreateReplacement(FilenameRange, "<" + FileName.str() + ">");
+  }
+
+  const std::string fileName(FileName);
+  const std::string project = fileName.substr(0, fileName.find_first_of('/'));
+  const std::string ext = fileName.substr(fileName.find_last_of('.'));
+  if (ext != ".hh" && ext != ".h") {
+    Check.diag(FilenameRange.getBegin(), "Only header files (.h, .hh) should be included");
+  }
+
+  const std::string currentDir = SM.getFileManager().getCanonicalName(SM.getFileEntryForID(SM.getFileID(HashLoc))->getDir()).str();
+  const std::string includeDir = SM.getFileManager().getCanonicalName(File->getDir()).str();
+
+  const size_t prefixSize = includeDir.find(project) + project.length();
+
+  GroupType groupType = GroupType::External;
+
+  if (currentDir == includeDir) {
+    groupType = GroupType::SameDirectory;
+
+    const std::string includeBasename = fileName.substr(fileName.find_last_of('/')).substr(1);
+    const std::string curFileName = SM.getFileEntryForID(SM.getFileID(HashLoc))->getName().str();
+    std::string curBasename = curFileName.substr(curFileName.find_last_of('/')).substr(1);
+
+    const std::string curExt = curFileName.substr(curFileName.find_last_of('.'));
```

Comment:
same concern as above. It might make sense to define some utility functions to extract extension and first/last component of the path.